### PR TITLE
Add functionality to get potential meeting times

### PR DESCRIPTION
**Pull request by AymarMuhikira**
-Separate events based on optional and requested attendees
-Find Collection of meeting times for requested attendees
-Check if optional attendees can attend any of those
-If so, return meeting times with optional attendees
-Otherwise, return meeting times for requested attendees only,

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,97 @@
...
+  }
+
+  private Collection<TimeRange> getOptionalTime(Collection<Event> events, long requestDuration, Collection<TimeRange> requiredTimes) {
+    return getTime(events, requestDuration, requiredTimes);
+  }
+
+  private int eventType(Event event, MeetingRequest request) {
+    HashSet<String> attendees = new HashSet<> (event.getAttendees());
+    attendees.retainAll(request.getAttendees());  
+    if (attendees.size() > 0) {
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Consider using `Collections.disjoint(a, b)` which tells you if `a` and `b` have any elements in common. You also don't need to turn the collections of attendees into hashsets before using disjoint.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,97 @@
...
+    }
+
+    return getTime(events, requestDuration, meetingTimes);
+  }
+
+  private Collection<TimeRange> getOptionalTime(Collection<Event> events, long requestDuration, Collection<TimeRange> requiredTimes) {
+    return getTime(events, requestDuration, requiredTimes);
+  }
+
+  private int eventType(Event event, MeetingRequest request) {
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Using arbitrary integer values to return a specific label is highly discouraged. It's better to create an `enum` with the three labels that you are using.

Better yet, just have two separate helpers `hasRequiredAttendees` and `hasOptionalAttendees` which return simple boolean values. There is no benefit to checking for both properties in the same method.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,97 @@
...
+      }
+      else if (type == 2) {
+        optionalEvents.add(event);    
+      }
+    }
+
+    Collection<TimeRange> requiredTimes = getRequiredTime(requiredEvents, request.getDuration());
+    Collection<TimeRange> optionalTimes = getOptionalTime(optionalEvents, request.getDuration(), requiredTimes);
+
+    if(requiredEvents.size() == 0 || optionalTimes.size() > 0) {
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I am a little confused about the first part of this condition: `requiredEvents.size() == 0`. I assume this is handling the test case `optionalWithNoGaps`? When there are no required attendees and no times can accommodate the optional attendees?

You should add a brief comment explaining what you are trying to do here, because it wasn't obvious when reading just this method by itself.

Also, I am not sure this is exactly the logic that you want. If you have some required attendees, who have no events previously scheduled, and some optional attendees that have no available times, then you should return the entire day to accommodate the required ones. (You should add another test case to cover this scenario!)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,97 @@
...
+            temp.add(division);
+          }
+        }
+        else if (time.contains(event.getWhen().end())) {
+          TimeRange division = TimeRange.fromStartEnd(event.getWhen().end(), time.end(), false);
+          if (division.duration() >= requestDuration && division.end() <= TimeRange.END_OF_DAY + 1) {
+            temp.add(division);  
+          }
+        }
+        else if (!event.getWhen().contains(time)){
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:87]: I think you can just use `else` here, because you've already covered all other possible cases above.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,97 @@
...
+    return requiredTimes;
+  }
+
+  private Collection<TimeRange> getTime(Collection<Event> events, long requestDuration, Collection<TimeRange> initial) {
+    Collection<TimeRange> meetingTimes = initial;
+
+    for (Event event: events) {
+      List<TimeRange> temp = new ArrayList<>();
+      for (TimeRange time: meetingTimes) {
+        if (time.contains(event.getWhen())){
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It would be helpful to add some comment describing (at a high level) what each of these cases look like and what the resulting events look like.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,97 @@
...
+    Collection<TimeRange> requiredTimes = getRequiredTime(requiredEvents, request.getDuration());
+    Collection<TimeRange> optionalTimes = getOptionalTime(optionalEvents, request.getDuration(), requiredTimes);
+
+    if(requiredEvents.size() == 0 || optionalTimes.size() > 0) {
+      return optionalTimes;  
+    }
+    return requiredTimes;
+  }
+
+  private Collection<TimeRange> getTime(Collection<Event> events, long requestDuration, Collection<TimeRange> initial) {
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Right now your algorithm looks at all the potential meeting times for every existing event in the calendar. If you imagine a two timelines with scheduled events on one and potential meeting times on the other, a scheduled event doesn't affect most of the potential meeting times. Can you do something to avoid do all the extra work of looking at every potential meeting time?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,97 @@
...
+    Collection<TimeRange> meetingTimes = initial;
+
+    for (Event event: events) {
+      List<TimeRange> temp = new ArrayList<>();
+      for (TimeRange time: meetingTimes) {
+        if (time.contains(event.getWhen())){
+          TimeRange firstDivision = TimeRange.fromStartEnd(time.start(), event.getWhen().start(), false);
+          TimeRange secondDivision = TimeRange.fromStartEnd(event.getWhen().end(), time.end(), false);
+          if (firstDivision.duration() >= requestDuration && firstDivision.end() <= TimeRange.END_OF_DAY + 1) {
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: I don't understand why you compare less-than-or-equal to `TimeRange.END_OF_DAY + 1`. Does this mean events can extend by one time unit into the next day/

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 8]
```diff
@@ -121,6 +124,67 @@ public void everyAttendeeIsConsidered() {
...
 
+  @Test
+  public void everyAttendeeAllDayOptional() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times, and the optional person has an all day event.
+    //
+    // Events  :       |--A--|     |--B--|
+    //Optional :|--------------C-------------|
+    // Day     : |-----------------------------|
+    // Options : |--1--|     |--2--|     |--3--|
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:136]: The diagrams are really helpful in the test cases!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 9]
```diff
@@ -121,6 +124,67 @@ public void everyAttendeeIsConsidered() {
...
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void everyAttendeeMorningOptional() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times, and the optional person has an all day event.
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: The first part of the comment doesn't match the diagram and the actual test code.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 10]
```diff
@@ -270,5 +362,62 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void optionalWithGaps() {
+    // Have one person, but make it so that there is not enough room at any point in the day to
+    // have the meeting.
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Please update the first part of the comment.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 11]
```diff
@@ -270,5 +362,62 @@ public void notEnoughRoom() {
...
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES),
+        TimeRange.fromStartEnd(TIME_0915AM,TIME_1100AM, false));
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:391]: It's a little confusing that you specify one of the expected time ranges using `fromStartDuration` and the other using `fromStartEnd`. I think using `fromStartEnd` is easier to understand.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -270,5 +362,62 @@ public void notEnoughRoom() {
...
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void optionalWithNoGaps() {
+    // Have one person, but make it so that there is not enough room at any point in the day to
+    // have the meeting.
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Please update the first part of the comment.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -270,5 +362,62 @@ public void notEnoughRoom() {
...
+        new Event("Event 3", TimeRange.fromStartEnd(TIME_1100AM, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_15_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList();
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: This expected outcome is not consistent with instructions in the walk-through: https://github.com/googleinterns/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md#writing-more-tests

It's OK if you want to change the specification, because it's a weird case that would "make sense" either way. Just update the comment at the top of this method to explain how you interpreted this situation.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed this

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Used separate helpers. Thank you, I didn't know about the practice of not using integers representing labels.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The first part of the condition was based on me not understanding the requirements of the [last test case](https://github.com/googleinterns/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md#writing-more-tests).

Fixed this

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:87]: Using else could lead to adding a potential time that is completely contained within an event (I didn't check that option before). Added a comment.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Added comments to describe the test cases.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: We could avoid looking at every potential meeting time by using binarySearch to find the meeting time closest to the event. I added the method getTimeAlternative() to show at a high level how this algorithm would work (It does not yet return the expected results for all test cases, I am looking into this issue). This alternative method does not necessarily reduce the run time, as sorting is O(n log n)  while going through every potential meeting time is O(n)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: The TimeRange.end() is the open interval end. The open interval end of the day is 24:00, as opposed to END_OF_DAY, which is 23:59. Added a comment

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 8 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:136]: Thank you!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 9 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Changed the test description

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 10 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Updated the comment

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 11 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:391]: Changed to fromStartEnd

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 12 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Updated the comment

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 13 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I did not understand correctly the last test case. Updated the method to test the required behavior.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (antmar) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:87]: Ah yes, good catch!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -22,52 +22,68 @@
...
-      else if (type == 2) {
+      //Only check for optional attendees if the event has no required one
+      else if (hasOptionalAttendees(event, request)) {
         optionalEvents.add(event);    
       }
     }
 
-    Collection<TimeRange> requiredTimes = getRequiredTime(requiredEvents, request.getDuration());
-    Collection<TimeRange> optionalTimes = getOptionalTime(optionalEvents, request.getDuration(), requiredTimes);
+    List<TimeRange> requiredTimes = getRequiredTime(requiredEvents, request.getDuration());
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: Out of curiosity, what was the motivation to change from `Collection` to `List` everywhere?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -89,22 +105,126 @@ else if (!event.getWhen().contains(time)){
...
+  }
+
+  private boolean hasOptionalAttendees(Event event, MeetingRequest request) {
+    Collection<String> eventAttendees = event.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    boolean hasAttendees = !Collections.disjoint(eventAttendees, optionalAttendees);
+    return hasAttendees;
+  }
+
+  private List<TimeRange> getTimeAlternative(Collection<Event> events, long requestDuration, List<TimeRange> initial) {
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Please make a separate PR with the alternative implementation. I think this PR is close to done, and you should submit it first.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,217 @@
...
+    List<TimeRange> meetingTimes = initial;
+
+    for (Event event: events) {
+      Collections.sort(meetingTimes, TimeRange.ORDER_BY_START);
+      int upperIndex = Collections.binarySearch(meetingTimes, event.getWhen(), TimeRange.ORDER_BY_START);
+      if(upperIndex < meetingTimes.size()) {
+        TimeRange upperTime = meetingTimes.get(upperIndex);
+        //time:     |--------------|
+        //event:         |----|
+        //division: |----|    |----|
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thank you - diagrams are very helpful!

**Review by antmar — COMMENTED** [review]
Just to be clear - this is ready to submit, after you move getTimeAlternative to another PR.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 14 above)
Comment (AymarMuhikira) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: The sort and binarySearch methods called in getTimeAlternative do not work on `Collection`, but do on `List`

**Review by antmar — APPROVED** [review]
